src/shared/clients/defaultS3Client.ts

```diff
@@ -0,0 +1,411 @@
+     * The list of sub-paths is returned in the result set and can be used in subsequent calls.
+     *
+     * A consequence of the fact that folders don't exist is that folders and files are intermingled across all
+     * of the pages.
+     * It's not possible to retrieve an exhaustive list of all folders without traversing all of the pages.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    public async listObjects(request: ListObjectsRequest): Promise<ListObjectsResponse> {
+        const s3 = await this.createS3()
+
+        let response: ListObjectsV2Output
+        try {
+            response = await s3
+                .listObjectsV2({
+                    Bucket: request.bucketName,
+                    Delimiter: DEFAULT_DELIMITER,
+                    MaxKeys: DEFAULT_MAX_KEYS,
+                    Prefix: request.folderPath,
+                    ContinuationToken: request.continuationToken,
+                })
+                .promise()
+        } catch (e) {
+            throw new S3Error(`Failed to list objects for bucket ${request.bucketName}: ${e}`)
+        }
+
+        const files: File[] = _(response.Contents)
+            .reject(file => file.Key === request.folderPath)
+            .map(
+                file =>
+                    new DefaultFile({
+                        key: file.Key!,
+                        partitionId: this.partitionId,
+                        bucketName: request.bucketName,
+                        lastModified: file.LastModified,
+                        sizeBytes: file.Size,
+                    })
+            )
+            .value()
+
+        const folders: Folder[] = _(response.CommonPrefixes)
+            .map(prefix => prefix.Prefix)
+            .compact()
+            .map(path => new DefaultFolder({ path, partitionId: this.partitionId, bucketName: request.bucketName }))
+            .value()
+
+        return {
+            files,
+            folders,
+            continuationToken: response.NextContinuationToken,
+        }
+    }
+
+    /**
+     * Looks up the region for the given bucket
+     *
+     * Note that although there is an S3#GetBucketLocation API,
+     * this is the suggested method of obtaining the region.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    private async lookupRegion(bucketName: string, s3: S3): Promise<string> {
+        try {
+            const response = await s3.headBucket({ Bucket: bucketName }).promise()
+            return response.$response.httpResponse.headers[BUCKET_REGION_HEADER]
+        } catch (e) {
+            throw new S3Error(`Failed to find region for bucket ${bucketName}: ${e}`)
+        }
+    }
+}
+
+export interface FileStreams {
+    createReadStream(uri: vscode.Uri): stream.Readable
+    createWriteStream(uri: vscode.Uri): stream.Writable
+}
+
+export class DefaultBucket implements Bucket {
```

Comment:
This might be lack of familiarity with the VSCode codebase but why are we wrapping AWS SDK structs? this seems like a pain from a maintenance perspective and looked like it was a pain to write.

Reply:
I very much agree: https://github.com/aws/aws-toolkit-vscode/pull/1075#discussion_r421802483

Reply:
This was actually avoiding the pain of integrating with and testing an unwrapped version.

I think it would be a stronger argument to keep the original interfaces if they offered most of what we want. However that's not the case here.

*Here's an example:*

### S3 SDK

#### API Call

```typescript
export class S3 {
  // Should call `.promise()` on result
  public createBucket(params: S3.Types.CreateBucketRequest, /* unused */ callback?: (err: AWSError, data: S3.Types.CreateBucketOutput) => void): Request<S3.Types.CreateBucketOutput, AWSError>
}
````

#### Request 

```typescript
  export interface CreateBucketRequest {
    ACL?: BucketCannedACL; // unused
    Bucket: BucketName;
    CreateBucketConfiguration?: CreateBucketConfiguration;
    GrantFullControl?: GrantFullControl; // unused
    GrantRead?: GrantRead; // unused
    GrantReadACP?: GrantReadACP; // unused
    GrantWrite?: GrantWrite; // unused
    GrantWriteACP?: GrantWriteACP; // unused
    ObjectLockEnabledForBucket?: ObjectLockEnabledForBucket; // unused
  }

  export type BucketCannedACL = "private"|"public-read"|"public-read-write"|"authenticated-read"|string; // unused

  export type BucketName = string; 

  export interface CreateBucketConfiguration {
    LocationConstraint?: BucketLocationConstraint; // redundant, read from constructor
  }

  export type BucketLocationConstraint = "EU"|"eu-west-1"|"us-west-1"|"us-west-2"|"ap-south-1"|"ap-southeast-1"|"ap-southeast-2"|"ap-northeast-1"|"sa-east-1"|"cn-north-1"|"eu-central-1"|string;

  export type GrantFullControl = string; // unused
  export type GrantRead = string; // unused
  export type GrantReadACP = string; // unused
  export type GrantWrite = string; // unused
  export type GrantWriteACP = string; // unused
  export type ObjectLockEnabledForBucket = boolean; // unused
```

#### Response

```typescript
  export interface CreateBucketOutput {
    Location?: Location; // unused
  }

  export type Location = string; // unused
```

### S3 Wrapper

#### API Call

```typescript
export class DefaultS3Client {
    public async createBucket(request: CreateBucketRequest): Promise<CreateBucketResponse>
}
```

#### Request

```typescript
export interface CreateBucketRequest {
    readonly bucketName: string
}
```

#### Response

```typescript
export interface CreateBucketResponse {
    readonly bucket: Bucket
}

// Becomes the model
export interface Bucket {
    readonly name: string
    readonly region: string
    readonly arn: string // not present in SDK repsonse
}
```